File: Extensification/Numbers/Long/Conversion.vb

'    Extensification  Copyright (C) 2020-2021  EoflaOE
'
'    This file is part of Extensification
'
'    Extensification is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Extensification is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.Runtime.CompilerServices

Namespace LongExts
    ''' <summary>
    ''' Provides the 64-bit integer extensions related to conversion
    ''' </summary>
    Public Module Conversion

        ''' <summary>
        ''' Converts number to a human readable format with units used
        ''' </summary>
        ''' <param name="Num">Number</param>
        ''' <param name="Type">Measurement types</param>
        ''' <returns>A string containing the processed number and unit</returns>
        <Extension>
        Public Function ToHumanReadable(Num As Long, Type As HumanFormats) As String
            Dim HumanString As String = ""
            Dim HumanNum As Double
            Select Case Type
                Case HumanFormats.DataSize
                    If Num >= 1000000000000000000 Then 'EB
                        HumanNum = Num / 1000000000000000000
                        HumanString = HumanNum & " EB"
                    ElseIf Num >= 1000000000000000 Then 'PB
                        HumanNum = Num / 1000000000000000
                        HumanString = HumanNum & " PB"
                    ElseIf Num >= 1000000000000 Then 'TB
                        HumanNum = Num / 1000000000000
                        HumanString = HumanNum & " TB"
                    ElseIf Num >= 1000000000 Then 'GB
                        HumanNum = Num / 1000000000
                        HumanString = HumanNum & " GB"
                    ElseIf Num >= 1000000 Then 'MB
                        HumanNum = Num / 1000000
                        HumanString = HumanNum & " MB"
                    ElseIf Num >= 1000 Then 'KB
                        HumanNum = Num / 1000
                        HumanString = HumanNum & " KB"
                    Else 'bytes
                        HumanString = HumanNum & " bytes"
                    End If
                Case HumanFormats.MeasurementsImperial
                    If Num >= 63360 Then 'miles
                        HumanNum = Num / 63360
                        HumanString = HumanNum & " miles"
                    ElseIf Num >= 36 Then 'yards
                        HumanNum = Num / 36
                        HumanString = HumanNum & " yards"
                    ElseIf Num >= 12 Then 'feet
                        HumanNum = Num / 12
                        HumanString = HumanNum & " feet"
                    Else 'inches
                        HumanString = Num & " inches"
                    End If
                Case HumanFormats.MeasurementsMetric
                    If Num >= 1000000 Then 'km
                        HumanNum = Num / 1000000
                        HumanString = HumanNum & " kilometers"
                    ElseIf Num >= 1000 Then 'm
                        HumanNum = Num / 1000
                        HumanString = HumanNum & " meters"
                    ElseIf Num >= 10 Then 'cm
                        HumanNum = Num / 10
                        HumanString = HumanNum & " centimeters"
                    Else 'mm
                        HumanString = Num & " millimeters"
                    End If
                Case HumanFormats.MeasurementsMetricUnusual
                    If Num >= 1000000 Then 'km
                        HumanNum = Num / 1000000
                        HumanString = HumanNum & " kilometers"
                    ElseIf Num >= 100000 Then 'hm
                        HumanNum = Num / 100000
                        HumanString = HumanNum & " hectameters"
                    ElseIf Num >= 10000 Then 'dcm
                        HumanNum = Num / 10000
                        HumanString = HumanNum & " decameters"
                    ElseIf Num >= 1000 Then 'm
                        HumanNum = Num / 1000
                        HumanString = HumanNum & " meters"
                    ElseIf Num >= 100 Then 'dm
                        HumanNum = Num / 100
                        HumanString = HumanNum & " decimeters"
                    ElseIf Num >= 10 Then 'cm
                        HumanNum = Num / 10
                        HumanString = HumanNum & " centimeters"
                    Else 'mm
                        HumanString = Num & " millimeters"
                    End If
                Case HumanFormats.VolumeMetric
                    If Num >= 1000000 Then 'kL
                        HumanNum = Num / 1000000
                        HumanString = HumanNum & " kiloliters"
                    ElseIf Num >= 1000 Then 'L
                        HumanNum = Num / 1000
                        HumanString = HumanNum & " liters"
                    Else 'mL
                        HumanString = Num & " milliliters"
                    End If
                Case HumanFormats.VolumeImperial
                    If Num >= 8 Then 'gallons
                        HumanNum = Num / 8
                        HumanString = HumanNum & " gallons"
                    ElseIf Num >= 2 Then 'quarts
                        HumanNum = Num / 2
                        HumanString = HumanNum & " quarts"
                    Else 'pints
                        HumanString = Num & " pints"
                    End If
            End Select
            Return HumanString
        End Function

        ''' <summary>
        ''' Converts number to a human readable format with units used
        ''' </summary>
        ''' <param name="Num">Number</param>
        ''' <param name="Type">Measurement types</param>
        ''' <returns>A string containing the processed number and unit</returns>
        <Extension>
        Public Function ToHumanReadable(Num As ULong, Type As HumanFormats) As String
            Dim HumanString As String = ""
            Dim HumanNum As Double
            Select Case Type
                Case HumanFormats.DataSize
                    If Num >= 1000000000000000000 Then 'EB
                        HumanNum = Num / 1000000000000000000
                        HumanString = HumanNum & " EB"
                    ElseIf Num >= 1000000000000000 Then 'PB
                        HumanNum = Num / 1000000000000000
                        HumanString = HumanNum & " PB"
                    ElseIf Num >= 1000000000000 Then 'TB
                        HumanNum = Num / 1000000000000
                        HumanString = HumanNum & " TB"
                    ElseIf Num >= 1000000000 Then 'GB
                        HumanNum = Num / 1000000000
                        HumanString = HumanNum & " GB"
                    ElseIf Num >= 1000000 Then 'MB
                        HumanNum = Num / 1000000
                        HumanString = HumanNum & " MB"
                    ElseIf Num >= 1000 Then 'KB
                        HumanNum = Num / 1000
                        HumanString = HumanNum & " KB"
                    Else 'bytes
                        HumanString = HumanNum & " bytes"
                    End If
                Case HumanFormats.MeasurementsImperial
                    If Num >= 63360 Then 'miles
                        HumanNum = Num / 63360
                        HumanString = HumanNum & " miles"
                    ElseIf Num >= 36 Then 'yards
                        HumanNum = Num / 36
                        HumanString = HumanNum & " yards"
                    ElseIf Num >= 12 Then 'feet
                        HumanNum = Num / 12
                        HumanString = HumanNum & " feet"
                    Else 'inches
                        HumanString = Num & " inches"
                    End If
                Case HumanFormats.MeasurementsMetric
                    If Num >= 1000000 Then 'km
                        HumanNum = Num / 1000000
                        HumanString = HumanNum & " kilometers"
                    ElseIf Num >= 1000 Then 'm
                        HumanNum = Num / 1000
                        HumanString = HumanNum & " meters"
                    ElseIf Num >= 10 Then 'cm
                        HumanNum = Num / 10
                        HumanString = HumanNum & " centimeters"
                    Else 'mm
                        HumanString = Num & " millimeters"
                    End If
                Case HumanFormats.MeasurementsMetricUnusual
                    If Num >= 1000000 Then 'km
                        HumanNum = Num / 1000000
                        HumanString = HumanNum & " kilometers"
                    ElseIf Num >= 100000 Then 'hm
                        HumanNum = Num / 100000
                        HumanString = HumanNum & " hectameters"
                    ElseIf Num >= 10000 Then 'dcm
                        HumanNum = Num / 10000
                        HumanString = HumanNum & " decameters"
                    ElseIf Num >= 1000 Then 'm
                        HumanNum = Num / 1000
                        HumanString = HumanNum & " meters"
                    ElseIf Num >= 100 Then 'dm
                        HumanNum = Num / 100
                        HumanString = HumanNum & " decimeters"
                    ElseIf Num >= 10 Then 'cm
                        HumanNum = Num / 10
                        HumanString = HumanNum & " centimeters"
                    Else 'mm
                        HumanString = Num & " millimeters"
                    End If
                Case HumanFormats.VolumeMetric
                    If Num >= 1000000 Then 'kL
                        HumanNum = Num / 1000000
                        HumanString = HumanNum & " kiloliters"
                    ElseIf Num >= 1000 Then 'L
                        HumanNum = Num / 1000
                        HumanString = HumanNum & " liters"
                    Else 'mL
                        HumanString = Num & " milliliters"
                    End If
                Case HumanFormats.VolumeImperial
                    If Num >= 8 Then 'gallons
                        HumanNum = Num / 8
                        HumanString = HumanNum & " gallons"
                    ElseIf Num >= 2 Then 'quarts
                        HumanNum = Num / 2
                        HumanString = HumanNum & " quarts"
                    Else 'pints
                        HumanString = Num & " pints"
                    End If
            End Select
            Return HumanString
        End Function

    End Module
End Namespace

File: Extensification/Numbers/Long/Conversion.cs

namespace Extensification.LongExts
{
    /// <summary>
    /// Provides the 64-bit integer extensions related to conversion
    /// </summary>
    public static class Conversion
    {

        /// <summary>
        /// Converts number to a human readable format with units used
        /// </summary>
        /// <param name="Num">Number</param>
        /// <param name="Type">Measurement types</param>
        /// <returns>A string containing the processed number and unit</returns>
        public static string ToHumanReadable(this long Num, HumanFormats Type)
        {
            string HumanString = "";
            var HumanNum = default(double);
            switch (Type)
            {
                case HumanFormats.DataSize:
                    {
                        if (Num >= 1000000000000000000L) // EB
                        {
                            HumanNum = Num / 1000000000000000000d;
                            HumanString = HumanNum + " EB";
                        }
                        else if (Num >= 1000000000000000L) // PB
                        {
                            HumanNum = Num / 1000000000000000d;
                            HumanString = HumanNum + " PB";
                        }
                        else if (Num >= 1000000000000L) // TB
                        {
                            HumanNum = Num / 1000000000000d;
                            HumanString = HumanNum + " TB";
                        }
                        else if (Num >= 1000000000L) // GB
                        {
                            HumanNum = Num / 1000000000d;
                            HumanString = HumanNum + " GB";
                        }
                        else if (Num >= 1000000L) // MB
                        {
                            HumanNum = Num / 1000000d;
                            HumanString = HumanNum + " MB";
                        }
                        else if (Num >= 1000L) // KB
                        {
                            HumanNum = Num / 1000d;
                            HumanString = HumanNum + " KB";
                        }
                        else // bytes
                        {
                            HumanString = HumanNum + " bytes";
                        }

                        break;
                    }
                case HumanFormats.MeasurementsImperial:
                    {
                        if (Num >= 63360L) // miles
                        {
                            HumanNum = Num / 63360d;
                            HumanString = HumanNum + " miles";
                        }
                        else if (Num >= 36L) // yards
                        {
                            HumanNum = Num / 36d;
                            HumanString = HumanNum + " yards";
                        }
                        else if (Num >= 12L) // feet
                        {
                            HumanNum = Num / 12d;
                            HumanString = HumanNum + " feet";
                        }
                        else // inches
                        {
                            HumanString = Num + " inches";
                        }

                        break;
                    }
                case HumanFormats.MeasurementsMetric:
                    {
                        if (Num >= 1000000L) // km
                        {
                            HumanNum = Num / 1000000d;
                            HumanString = HumanNum + " kilometers";
                        }
                        else if (Num >= 1000L) // m
                        {
                            HumanNum = Num / 1000d;
                            HumanString = HumanNum + " meters";
                        }
                        else if (Num >= 10L) // cm
                        {
                            HumanNum = Num / 10d;
                            HumanString = HumanNum + " centimeters";
                        }
                        else // mm
                        {
                            HumanString = Num + " millimeters";
                        }

                        break;
                    }
                case HumanFormats.MeasurementsMetricUnusual:
                    {
                        if (Num >= 1000000L) // km
                        {
                            HumanNum = Num / 1000000d;
                            HumanString = HumanNum + " kilometers";
                        }
                        else if (Num >= 100000L) // hm
                        {
                            HumanNum = Num / 100000d;
                            HumanString = HumanNum + " hectameters";
                        }
                        else if (Num >= 10000L) // dcm
                        {
                            HumanNum = Num / 10000d;
                            HumanString = HumanNum + " decameters";
                        }
                        else if (Num >= 1000L) // m
                        {
                            HumanNum = Num / 1000d;
                            HumanString = HumanNum + " meters";
                        }
                        else if (Num >= 100L) // dm
                        {
                            HumanNum = Num / 100d;
                            HumanString = HumanNum + " decimeters";
                        }
                        else if (Num >= 10L) // cm
                        {
                            HumanNum = Num / 10d;
                            HumanString = HumanNum + " centimeters";
                        }
                        else // mm
                        {
                            HumanString = Num + " millimeters";
                        }

                        break;
                    }
                case HumanFormats.VolumeMetric:
                    {
                        if (Num >= 1000000L) // kL
                        {
                            HumanNum = Num / 1000000d;
                            HumanString = HumanNum + " kiloliters";
                        }
                        else if (Num >= 1000L) // L
                        {
                            HumanNum = Num / 1000d;
                            HumanString = HumanNum + " liters";
                        }
                        else // mL
                        {
                            HumanString = Num + " milliliters";
                        }

                        break;
                    }
                case HumanFormats.VolumeImperial:
                    {
                        if (Num >= 8L) // gallons
                        {
                            HumanNum = Num / 8d;
                            HumanString = HumanNum + " gallons";
                        }
                        else if (Num >= 2L) // quarts
                        {
                            HumanNum = Num / 2d;
                            HumanString = HumanNum + " quarts";
                        }
                        else // pints
                        {
                            HumanString = Num + " pints";
                        }

                        break;
                    }
            }
            return HumanString;
        }

        /// <summary>
        /// Converts number to a human readable format with units used
        /// </summary>
        /// <param name="Num">Number</param>
        /// <param name="Type">Measurement types</param>
        /// <returns>A string containing the processed number and unit</returns>
        public static string ToHumanReadable(this ulong Num, HumanFormats Type)
        {
            string HumanString = "";
            var HumanNum = default(double);
            switch (Type)
            {
                case HumanFormats.DataSize:
                    {
                        if (Num >= 1000000000000000000m) // EB
                        {
                            HumanNum = Num / 1000000000000000000d;
                            HumanString = HumanNum + " EB";
                        }
                        else if (Num >= 1000000000000000m) // PB
                        {
                            HumanNum = Num / 1000000000000000d;
                            HumanString = HumanNum + " PB";
                        }
                        else if (Num >= 1000000000000m) // TB
                        {
                            HumanNum = Num / 1000000000000d;
                            HumanString = HumanNum + " TB";
                        }
                        else if (Num >= 1000000000m) // GB
                        {
                            HumanNum = Num / 1000000000d;
                            HumanString = HumanNum + " GB";
                        }
                        else if (Num >= 1000000m) // MB
                        {
                            HumanNum = Num / 1000000d;
                            HumanString = HumanNum + " MB";
                        }
                        else if (Num >= 1000m) // KB
                        {
                            HumanNum = Num / 1000d;
                            HumanString = HumanNum + " KB";
                        }
                        else // bytes
                        {
                            HumanString = HumanNum + " bytes";
                        }

                        break;
                    }
                case HumanFormats.MeasurementsImperial:
                    {
                        if (Num >= 63360m) // miles
                        {
                            HumanNum = Num / 63360d;
                            HumanString = HumanNum + " miles";
                        }
                        else if (Num >= 36m) // yards
                        {
                            HumanNum = Num / 36d;
                            HumanString = HumanNum + " yards";
                        }
                        else if (Num >= 12m) // feet
                        {
                            HumanNum = Num / 12d;
                            HumanString = HumanNum + " feet";
                        }
                        else // inches
                        {
                            HumanString = Num + " inches";
                        }

                        break;
                    }
                case HumanFormats.MeasurementsMetric:
                    {
                        if (Num >= 1000000m) // km
                        {
                            HumanNum = Num / 1000000d;
                            HumanString = HumanNum + " kilometers";
                        }
                        else if (Num >= 1000m) // m
                        {
                            HumanNum = Num / 1000d;
                            HumanString = HumanNum + " meters";
                        }
                        else if (Num >= 10m) // cm
                        {
                            HumanNum = Num / 10d;
                            HumanString = HumanNum + " centimeters";
                        }
                        else // mm
                        {
                            HumanString = Num + " millimeters";
                        }

                        break;
                    }
                case HumanFormats.MeasurementsMetricUnusual:
                    {
                        if (Num >= 1000000m) // km
                        {
                            HumanNum = Num / 1000000d;
                            HumanString = HumanNum + " kilometers";
                        }
                        else if (Num >= 100000m) // hm
                        {
                            HumanNum = Num / 100000d;
                            HumanString = HumanNum + " hectameters";
                        }
                        else if (Num >= 10000m) // dcm
                        {
                            HumanNum = Num / 10000d;
                            HumanString = HumanNum + " decameters";
                        }
                        else if (Num >= 1000m) // m
                        {
                            HumanNum = Num / 1000d;
                            HumanString = HumanNum + " meters";
                        }
                        else if (Num >= 100m) // dm
                        {
                            HumanNum = Num / 100d;
                            HumanString = HumanNum + " decimeters";
                        }
                        else if (Num >= 10m) // cm
                        {
                            HumanNum = Num / 10d;
                            HumanString = HumanNum + " centimeters";
                        }
                        else // mm
                        {
                            HumanString = Num + " millimeters";
                        }

                        break;
                    }
                case HumanFormats.VolumeMetric:
                    {
                        if (Num >= 1000000m) // kL
                        {
                            HumanNum = Num / 1000000d;
                            HumanString = HumanNum + " kiloliters";
                        }
                        else if (Num >= 1000m) // L
                        {
                            HumanNum = Num / 1000d;
                            HumanString = HumanNum + " liters";
                        }
                        else // mL
                        {
                            HumanString = Num + " milliliters";
                        }

                        break;
                    }
                case HumanFormats.VolumeImperial:
                    {
                        if (Num >= 8m) // gallons
                        {
                            HumanNum = Num / 8d;
                            HumanString = HumanNum + " gallons";
                        }
                        else if (Num >= 2m) // quarts
                        {
                            HumanNum = Num / 2d;
                            HumanString = HumanNum + " quarts";
                        }
                        else // pints
                        {
                            HumanString = Num + " pints";
                        }

                        break;
                    }
            }
            return HumanString;
        }

    }
}